src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/seed/cas1/ApprovedPremisesRoomsSeedFromXLSXJob.kt

```diff
@@ -0,0 +1,210 @@
+      val rowId = dataFrame.getColumn(0).values().indexOf(question)
+
+      if (rowId == -1) throw SiteSurveyImportException("Characteristic question '$question' not found on sheet Sheet3.")
+
+      for (colId in 1..<dataFrame.columnsCount()) {
+        val roomCode = dataFrame.getColumn(colId).name
+        val answer = dataFrame[rowId][colId].toString().trim()
+
+        if (answer.equals("yes", true)) {
+          premisesCharacteristics.computeIfAbsent(roomCode) { mutableListOf() }.add(characteristic!!)
+        } else if (!answer.equals("no", true) && !answer.equals("N/A", true)) {
+          throw SiteSurveyImportException("Expecting 'yes' or 'no' for question '$question' but is '$answer' on sheet Sheet3.")
+        }
+      }
+    }
+    return premisesCharacteristics
+  }
+
+  private fun buildBeds(dataFrame: DataFrame<*>, rooms: MutableList<RoomEntity>): List<BedEntity> {
+    val beds = mutableListOf<BedEntity>()
+    for (i in 1..<dataFrame.columnsCount()) {
+      val roomAnswers = dataFrame.getColumn(i)
+      val bedNumber = roomAnswers[1].toString()
+      beds.add(
+        createBed(
+          "${roomAnswers.name}",
+          "${roomAnswers.name} - $bedNumber",
+          rooms.firstOrNull { it.code == roomAnswers.name }
+            ?: throw IllegalArgumentException("Room not found with id ${roomAnswers.name} and code ${roomAnswers.name}."),
+        ),
+      )
+    }
+    return beds
+  }
+
+  private fun createOrUpdateRooms(rooms: MutableList<RoomEntity>): MutableList<RoomEntity> {
+    rooms.forEachIndexed { index, room ->
+      var persistedRoom = roomRepository.findByCodeAndPremisesId(room.code!!, room.premises.id)
+
+      if (persistedRoom == null) {
+        roomRepository.save(room)
+        log.info("Created new room ${room.id} with code ${room.code} and name ${room.name} in premise ${room.premises.name}.")
+        log.info("Added characteristic(s) ${room.characteristics.joinToString()} to room code ${room.code}.")
+      } else {
+        rooms[index] = persistedRoom
+        persistedRoom.characteristics.clear()
+        persistedRoom.characteristics.addAll(room.characteristics)
+        roomRepository.save(persistedRoom)
+        log.info("Added characteristic(s) ${persistedRoom.characteristics.joinToString()} to room code ${persistedRoom.code}.")
+      }
+    }
+    return rooms
+  }
+
+  private fun createBedsIfNotExist(beds: List<BedEntity>) {
+    beds.forEach {
+      val existingBed = bedRepository.findByCodeAndRoomId(it.code!!, it.room.id)
+      if (existingBed != null) {
+        log.info("Bed ${it.id} with code ${it.code} already exists in room code ${it.room.code}.")
+      } else {
+        bedRepository.save(it)
+        log.info("Created new bed ${it.id} with code ${it.code} and name ${it.name} in room code ${it.room.code}.")
+      }
+    }
+  }
+
+  private fun createRoom(premisesId: UUID, roomCode: String, roomName: String): RoomEntity = RoomEntity(
```

Comment:
i think for consistency with code earlier on this should be called `buildRoom` (as we're not actually creating in the database at this point?)

Reply:
Renamed to `buildRoom`.